Computing roots of polynomials with real algebraic coefficients

mmaaz-git opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Addresses a challenge brought up in the discussion in #26785. #23077 was also previously opened to do the same thing but was not merged. This was initially brought up in #22943 I believe.

#### Brief description of what is fixed or changed

We can now compute roots of polynomials with algebraic coefficients. This would previously fail. It computes the "rational lift" of the polynomial, i.e., a polynomial with rational coefficients whose roots are a superset of the original polynomial's, and then uses numerical comparison to find which roots are also roots of the original polynomial. This leverages the recent fixes in #26813 and #26816. The numerical comparison code is taken nearly verbatim from https://github.com/sympy/sympy/issues/22943#issuecomment-2220562152.

Examples:
```
>>> from sympy import *
>>> from sympy.abc import x,y,z
>>> Poly(x**3 + sqrt(2)*x**2 - 1, x).real_roots()
[CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 0)]
>>> Poly(x**3 + sqrt(2)*x**2 - 1, x).all_roots()
[CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 0), CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 2), CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 3)]
>>> Poly(x**3 + CRootOf(y**5+y**3-1,0)*x**2 - 1, x).real_roots()
[CRootOf(x**15 + x**13 - 5*x**12 - 2*x**10 + 10*x**9 + 3*x**7 - 10*x**6 - x**4 + 5*x**3 - 1, 0)]
```

The fix is in ComplexRootOf._get_roots(), which is ultimately what Poly.real_roots() or Poly.all_roots() call (and hence ultimately what functions like solve() call I think). I think this makes logical sense in terms of integrating well into how the root functions already work. But, happy to hear feedback on this.

#### Other comments

This change breaks a test in `test_quintics_3()` in solvers/tests/test_solvers.py. The test asserted:

```
y = x**5 + x**3 - 2**Rational(1, 3)
assert solve(y) == solve(-y) == []
```

But now that this polynomial can be solved the roots should not be an empty list, but `solve(y) == solve(-y)` should still be correct. I changed the test accordingly. As well, the computation for complex roots can be somewhat slow, eg this test takes 210 seconds on my computer, hence I added a slow decorator.


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* polys
  * It is now possible to compute the roots of polynomials with algebraic coefficients as RootOf using `real_roots(p, extension=True)` or `all_roots(p, extension=True)`.

<!-- END RELEASE NOTES -->

mmaaz-git committed:
rootof: Finding roots of polynomials with real algebraic coefficients

Previously, SymPy could not solve for the roots of polynomials with real
algebraic coefficients, e.g. radicals or root objects. This change computes
the "rational lift" of the polynomial, i.e., a polynomial with rational
coefficients whose roots are a superset of the original polynomial's, and
then numerically filters them.

Example:

Poly(x**3 + sqrt(2)*x**2 - 1, x).real_roots() == [
        rootof(x**6 - 2*x**4 - 2*x**3 + 1, 0)
]

Previously, this would fail.

It uses the new functionality in #26813 and #26816 for constructing the
extension and correct root counting. As well, it adapts code from #22943.

mmaaz-git committed:
Removed prints

mmaaz-git committed:
Updated mailmap

mmaaz-git committed:
Updated mailmap

mmaaz-git committed:
Modified polytools doctests

mmaaz-git commented:
Had to change the 'real_roots()` and `all_roots()` doctests in polys/polytools.py as they previously expected a not implemented exception

smichr commented on sympy/polys/polytools.py:
```diff
@@ -6744,17 +6744,15 @@ def all_roots(f, multiple=True, radicals=True):
     [1.17, -0.765 - 0.352*I, -0.765 + 0.352*I, 0.181 - 1.08*I, 0.181 + 1.08*I]
 
-    Irrational algebraic or transcendental coefficients cannot currently be
+    Irrational algebraic or transcendental coefficients can now be
```
```suggestion
    Irrational algebraic or transcendental coefficients are
```

smichr commented on sympy/polys/polytools.py:
```diff
@@ -6920,17 +6918,15 @@ def real_roots(f, multiple=True, radicals=True):
     than real roots.
 
-    Irrational algebraic or transcendental coefficients cannot be handled by
+    Irrational algebraic or transcendental coefficients can now be handled by
```
```suggestion
    Irrational algebraic or transcendental coefficients are handled by
```

mmaaz-git committed:
Added check to make sure all coeffs are *real* algebraic before lifting

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -774,13 +774,58 @@ def _get_roots(cls, method, poly, radicals):
                 poly = poly.xreplace({d: x})
                 break
+
+        # compute rational lift if real algebraic
+        has_alg_coeff = any(c.is_algebraic and not c.is_rational
+                            for c in poly.all_coeffs())
+        has_alg_coeff = has_alg_coeff and all(c.is_real
+                            for c in poly.all_coeffs())
```
Check `is_algebraic` is not quite right because:
```python
In [1]: x = Symbol('x', integer=True)

In [2]: x
Out[2]: x

In [3]: x.is_algebraic
Out[3]: True
```
The check should really be `c.is_algebraic and c.is_number`. However I don't think we need to do this check because the `Poly` constructor already handles this.

mmaaz-git replied:
Ah good point

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -774,13 +774,58 @@ def _get_roots(cls, method, poly, radicals):
+        if has_alg_coeff:
+            poly_original = poly
+            poly = Poly(poly.expr, poly.gen, extension=True)
+            poly = poly.lift()
```
There needs to be a way to prevent `extension = True` being used because it can be very slow. Sometimes it would be better to get an exception here rather than attempting to compute a primitive element.

The usual way of handling this would be to say that `real_roots` should accept `**kwargs` and then pass those through to the `Poly` constructor. Then a user would have to do `real_roots(..., extension=True)`. I wonder though whether for this method in particular it makes sense to have `extension = True` as the default though.

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -774,13 +774,58 @@ def _get_roots(cls, method, poly, radicals):
             roots.append(coeff*cls._postprocess_root(root, radicals))
+
+        # filter out extraneous roots if introduced from lifting
+        if has_alg_coeff:
```
Rather than checking this in multiple places throughout the function it would be better to have two methods `_get_roots_qq` and `_get_roots_alg`. Then dispatch on `poly.domain`:
```
if poly.domain.is_QQ:
    return cls._get_roots_qq(...)
elif poly.domain.is_AlgebraicField:
    return cls._get_roots_alg(...)
```

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -774,13 +774,58 @@ def _get_roots(cls, method, poly, radicals):
+        else:
+            num_roots = poly_alg.degree()
+
+        prec = 10
+        # compare len(set()) bc expected behavior is for multiple roots
+        while len(set(candidates)) > num_roots:
+            candidates_filtered = []
+            for c in candidates:
+                r_f = poly_alg(c).evalf(prec, maxn=2*prec)
+                if abs(r_f)._prec < 2: candidates_filtered.append(c)
+            prec *= 2
+            candidates = candidates_filtered
```
If there are multiple roots then the inner loop should not repeat the calculation for them.

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -329,10 +329,10 @@ def test_quintics_2():
 
-
+@slow
 def test_quintics_3():
     y = x**5 + x**3 - 2**Rational(1, 3)
-    assert solve(y) == solve(-y) == []
+    assert solve(y) == solve(-y)
```
Presumably the change to `real_roots` affects something in `solve`. We need to consider whether that change is wanted. It might be better to disable that change for now by using `real_roots(..., extension=False)` or something.

oscarbenjamin replied:
Or otherwise this test should be updated to show what is actually returned.

mmaaz-git replied:
I agree with checking what is actually returned. I will add this in.

oscarbenjamin replied:
I'm wondering where it is in solve that something is effected by the change in this PR and whether is is a good thing that it is affected. I assume that `all_roots` is being used somewhere but it might be better to use extension=False or something.

mmaaz-git replied:
Yup `solve()` calls `poly.all_roots()`. I'll get back to this test after how `real_roots()` and `all_roots()` are called is finalized

mmaaz-git replied:
I've changed this test back as extension=False is default, so solve indeed returns an empty list.

oscarbenjamin commented:
There seem to be bugs somewhere because these cases should succeed:
```python
In [12]: p = Poly(x**5 + x**3 - 2**Rational(1, 3), x, extension=True)

In [13]: p
Out[13]: Poly(x**5 + x**3 - 2**(1/3), x, domain='QQ<2**(1/3)>')

In [14]: p.lift()
---------------------------------------------------------------------------
CoercionFailed 
...
CoercionFailed: Cannot convert ANP([-1, 0, 0], [1, 0, 0, -2], QQ) of type <class 'sympy.polys.polyclasses.ANP'> from QQ<2**(1/3)> to QQ
```
Also
```python
In [15]: Poly(x**2 + I).lift()
---------------------------------------------------------------------------
DomainError
...
DomainError: computation can be done only in an algebraic domain
```

mmaaz-git committed:
Update doctest

Co-authored-by: Christopher Smith <smichr@gmail.com>

mmaaz-git committed:
Update doctest

Co-authored-by: Christopher Smith <smichr@gmail.com>

mmaaz-git committed:
Updated mailmap

mmaaz-git committed:
rootoftools: Modify when algebraic extension is constructed

In the previous commits, when solving for roots of polynomial
with algebraic coeffs, both rational and algebraic coeffs were
handled in the function _get_root(). Now, _get_root() checks
for the domain and then dispatches to the two new functions
_get_root_qq() or _get_root_alg(). The former is basically the
same as the master branch, the latter includes the lifting and
numerical filtering procedures. As well, it now assumes that the
polynomial already has an algebraic extension field. This way,
whether or not the extension is constructed can be handled "higher
up" in the chain.

Lastly, in _numerically_filter_roots(), it checks whether the same
root has already been processed to not loop over them again.

mmaaz-git committed:
Fixing conversion of alg field to EX

mmaaz-git commented:
I've changed the implementation. There are now two functions `_get_roots_qq()` and `_get_roots_alg()`. It performs a check and dispatches as suggested. As well, the extension is not constructed within CRootOf anymore, the responsibility is punted "higher up", so that it can be turned off. Lastly, I've modified the loop for `_numerically_filter_roots` so that it keeps track of if it's already processed the root, in the case of duplicates. 

Since the extension has to be set before calling the root, I changed the `real_roots()` and `all_roots()` functions to set `extension=True`. This conflicted with the current `greedy` setting so I removed it for now. I'm not sure if this breaks anything else. Also, the unit tests now specify `extension=True`. 

Basically, this won't work:
```
Poly(sqrt(2)*x**2 - 1, x).real_roots()
```

But this will:
```
Poly(sqrt(2)*x**2 - 1, x, extension=True).real_roots()
```

Not sure if you think this is good. I'm not sure what the best way is for users to interact with this.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -7003,7 +6999,7 @@ def real_roots(f, multiple=True, radicals=True):
     """
     try:
-        F = Poly(f, greedy=False)
+        F = Poly(f, extension=True)
```
I think that there needs to be some way to control this when calling `real_roots` or `all_roots`. It might make sense for `extension=True` to be the default for these functions but then there at least needs to be a way to pass `extension=False` when calling the functions.

mmaaz-git replied:
Perhaps by adding a new argument? Currently, `real_roots()` or `all_roots()` take arguments `multiple` and `radical`. Or perhaps using `**kwargs`?

oscarbenjamin replied:
The obvious choice would be for them to take an argument `extension` that defaults to False. That would be in keeping with other Poly functions.

mmaaz-git replied:
Sure sounds good

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,94 @@ def _get_roots(cls, method, poly, radicals):
         free_names = {str(i) for i in poly.free_symbols}
         for x in chain((symbols('x'),), numbered_symbols('x')):
             if x.name not in free_names:
+                # xreplace makes alg field domain back into EX
+                # XXX: so make sure it stays same?
                 poly = poly.xreplace({d: x})
+                if dom.is_AlgebraicField:
+                    poly = Poly(poly, domain=dom)
```
Why does this need to change the symbols at all?

Is it possible to just remove this whole section or does that break something?

mmaaz-git replied:
I looked into this. Removing it breaks `test_issue_19113()` in test_rootoftools. This led me to #19113 and #19117 -- I think basically this symbol change is done to always return RootOf objects with canonical `x` generator.

mmaaz-git replied:
Changing the `xreplace` to a `sub` preserves the domain, i.e. all tests pass with `sub` and removing lines 781-782 above. But, not sure if there are any other differences between `xreplace` and `sub` we should be careful for.

oscarbenjamin replied:
There should be a better way than using `xreplace` that preserves the domain. It seems that `replace` can do it: `poly.replace(d, x)`.

oscarbenjamin replied:
The subs method just calls replace.

mmaaz-git replied:
Sounds good, thanks

oscarbenjamin replied:
See also gh-26808

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,94 @@ def _get_roots(cls, method, poly, radicals):
+                    poly = Poly(poly, domain=dom)
                 break
+
+        if dom.is_QQ or dom.is_ZZ:
+            return cls._get_roots_qq(method, poly, radicals)
+            # XXX: any better way to check if real?
+            # without this check, some doctests fail (with complex terms)
+        if dom.is_AlgebraicField and all(c.is_real for c in poly.coeffs()):
+            return cls._get_roots_alg(method, poly, radicals)
```
Why do we need to check `is_real` here?

Once we lift the polynomial the coefficients will be real.

mmaaz-git replied:
You're right, it's unnecessary.

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,94 @@ def _get_roots(cls, method, poly, radicals):
+            num_roots = poly.degree()
+
+        prec = 10
+        # compare len(set()) bc expected behavior is for multiple roots
+        while len(set(candidates)) > num_roots:
```
Is the multiple root case not handled earlier somehow?

I would have thought that by the time we get here the polynomial is irreducible and so cannot have multiple roots.

mmaaz-git replied:
Before the point `_numerically_filter_roots()` is called, the roots are created by `ComplexRootOf._real_roots()` or `ComplexRootOf._all_roots()` which return multiples.

Eg
```
>>> ComplexRootOf._real_roots(Poly((x-1)**2))
[(PurePoly(x - 1, x, domain='ZZ'), 0), (PurePoly(x - 1, x, domain='ZZ'), 0)]
```

oscarbenjamin replied:
That makes me think that we might be handling this in the wrong place. The RootOf code is a bit convoluted which makes it hard to follow through.

Potentially it is better to compute the square free factorisation of the polynomial first and then do the lift on the square free parts. Then we don't have multiple roots and we work with smaller polynomials.

I might be wrong though and maybe it is just best to do the lift at the outset so that all the other steps can use efficient calculations in `QQ`.

mmaaz-git replied:
My understanding of the RootOf code is (eg for real roots):

`_get_roots()` --> `_real_roots()` --> gets factors, calls `_get_reals()` --> `_get_reals_sqf()` which computes root isolating intervals for a square free factor

So it seems like the square free stuff is done further along currently. If doing the lift is better on smaller polynomials, then we could just change `_get_roots_alg()` to incorporate square free factorization before calling `_get_roots()`.

oscarbenjamin replied:
> If doing the lift is better on smaller polynomials, then we could just change `_get_roots_alg()` to incorporate square free factorization before calling `_real_roots()`.

I'm not sure that it is better to do this. Factorising over extension fields is also expensive so it might actually be better to lift the original polynomial so that all subsequent calculations are done with only polynomials having rational coefficients. Hard to say without more analysis/timings. For now I think this is fine if it works.

oscarbenjamin commented:
> Basically, this won't work:
> 
> ```
> Poly(sqrt(2)*x**2 - 1, x).real_roots()
> ```
> 
> But this will:
> 
> ```
> Poly(sqrt(2)*x**2 - 1, x, extension=True).real_roots()
> ```
> 
> Not sure if you think this is good.

I think that is good. I am not sure though that `real_roots(expr)` and `all_roots(expr)` should use `extension=True` by default though because using `extension=True` leads to computing a primitive element which can be expensive. If we already have a Poly with an Algebraicfield domain then the primitive element was already computed and then I think it should be fine to compute the lift.

mmaaz-git committed:
Remove check for all real coeffs

mmaaz-git committed:
polytools: New method to numerically filter roots

The ._numerically_filter_roots() method takes a set of candidate
roots that are a superset of f's roots and uses increasing
numerical precision to filter them out until we equal f's root
count, counted with multiplicity. This is necessary for getting roots
of a polynomial with algebraic coefficients from the roots of its
rational lift, and is right now only used for that purpose.

mmaaz-git committed:
Changed xreplace to replace to preserve domain

mmaaz-git committed:
Fixed replace call

mmaaz-git commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,94 @@ def _get_roots(cls, method, poly, radicals):
+        else:
+            # XXX: not sure how to handle ZZ[x] which appears in some tests?
+            # this makes the tests pass alright but has to be a better way?
+            return cls._get_roots_qq(method, poly, radicals)
```
One example is in integrals/tests/test_prde.py under test_param_rischDE, line 224. It seems it tries to compute the real roots of the following.
```
Poly((1 - x)*_z + 1 - x, _z, domain='ZZ[x]')
```
Without the `else` return, `None` is returned and then it throws a TypeError. 

This is reproducible (if the final `else:return` is removed) as:
```
>>> from sympy import *
>>> from sympy.abc import x,z
>>> print(Poly((1 - x)*z + 1 - x, z, domain='ZZ[x]').real_roots())
None
>>> 
```

But adding in the final `else: return cls._get_roots_qq()` fixes this. The correct answer should be that the root is -1.

oscarbenjamin replied:
> The correct answer should be that the root is -1.

I think the correct answer should be that `real_roots` rejects this polynomial because the domain is non-numeric.

See also gh-26787

mmaaz-git replied:
I'm uncertain about this because if the final `else: return` is removed, three tests break, the one above and two more.
They are:
- sympy/integrals/tests/test_prde.py:224: in test_param_rischDE
- sympy/integrals/tests/test_risch.py:206: in test_recognize_log_derivative
- sympy/series/tests/test_limitseq.py:81: in test_limit_seq

The polynomials in question are respectively:
- `Poly((1 - x)*_z + 1 - x, _z, domain='ZZ[x]')`
- `Poly((2*x**4 - 8*x**3 + 8*x**2)*z**2 - 2*x**4 + 8*x**3 - 8*x**2, z, domain='ZZ[x]')`
- `Poly(288*_p + 288*EulerGamma, _p, domain='ZZ[EulerGamma]')`

I'm not sure if `real_roots()` should reject these, or if these functions should somehow be changed so that the domain is numeric, or if they should be solved another way, perhaps with `ground_roots()`?

mmaaz-git commented:
I moved `_numerically_filter_roots()` to be a method for `Poly` as you suggested. The issue is, the way it works now, it only works if the candidates are known a priori to be a superset of the actual roots. Otherwise it does not work. 

Eg
```
>>> from sympy import *
>>> from sympy.abc import x,y,z
>>> Poly(x**2 - 1, x)._numerically_filter_roots([-10, 5], real=True)
[-10, 5]
```

Now that it is a method in `Poly`, should this be a more general method?

oscarbenjamin commented:
> The issue is, the way it works now, it only works if the candidates are known a priori to be a superset of the actual roots. Otherwise it does not work.

This is also true of `Poly.same_root`. I think that is fine as long as documented.

We could perhaps call this method `Poly.which_roots`?

mmaaz-git committed:
polytools: Changed _numerically_filter_roots to which_roots

Changed name of function. Added unit tests and doctests. Changed
references to it in rootoftools.

mmaaz-git committed:
polytools: New argument `extension` to real_roots() and all_roots()

Can now specify whether to construct the algebraic extension domain
when calling real_roots() or all_roots(). Defaults to false. Added
corresponding information to docstring.

mmaaz-git commented:
It has been changed to `which_roots()`. And an extension argument added to `all_roots()` and `real_roots()`. 

Now that this argument exists, I wonder if we should add something like that to `solve()`? Or what `solve()` should default to when it calls `all_roots()`?

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3816,72 @@ def nth_power_roots_poly(f, n):
+            num_roots = f.degree()
+
+        prec = 10
+        # compare len(set()) bc expected behavior is for multiple roots
+        while len(set(candidates)) > num_roots:
+            candidates_filtered = []
+            processed = set() # track already seen
+
+            for c in candidates:
+                if c not in processed:
+                    r_f = f(c).evalf(prec, maxn=2*prec)
+                    processed.add(c)
+                    if abs(r_f)._prec < 2:
+                        candidates_filtered.append(c)
+                else:
+                    if c in candidates_filtered:
+                        # already approved this candidate
+                        candidates_filtered.append(c)
```
I would write this as:
```python
root_counts = Counter(candidates)
roots_remaining = lambda: sum(root_counts.values())

while roots_remaining() > num_roots:
    for r in list(root_counts.keys()):
        # If f(r) != 0 then f(r).evalf() gives a float/complex with precision.
        f_r = f(r).evalf(prec, maxn=2*prec)
        if abs(f_r)._prec >= 2:
            root_counts.pop(c)
    prec *= 2

roots = sum([r]*c for r, c in root_counts.items(), [])
```

mmaaz-git replied:
I like this! I have changed the implementation to this. 

However, I found an unforeseen bug, which causes this implementation to loop indefinitely (until overflow of prec) -- and was also causing issues with the prior implementation. The issue is with multiplicity of real roots. This breaks real_roots() sometimes.

```
>>> Poly((x-1)**2, x).which_roots([1, 1])
OverflowError: cannot convert float infinity to integer

>>> Poly((x-sqrt(2))**2, x, extension=True).real_roots()
OverflowError: cannot convert float infinity to integer
```

The issue in the last example is that the lift has roots `[-sqrt(2), -sqrt(2), sqrt(2), sqrt(2)]`, the `-sqrt(2)` gets filtered out, but then roots_remaining() always stays 2, while num_roots is 1 -- this is because .count_roots() counts unique roots.

I think it is better to just compare the count of unique roots, but then add the multiplicity back in. This also leads to more consistent output between real=True and real=False (which uses degree to count roots).

I've added a function to Poly which is `root_multiplicity()` to accomplish this.

oscarbenjamin replied:
I guess it is not easily possible to have `count_roots` return a count with multiplicities.

I think this means that we need to factorise the polynomial first. We could simply require in this method that the polynomial be square-free. It is not necessary to check that it is square-free but we can document the requirement and make sure it is satisfied before calling the function.

This goes back to what I said before about this being handled in the wrong place. The multiple roots case should be handled at a higher-level using factorisation and then at some lower level the problem to be solved only involves square free polynomials and multiple roots do not need to be considered.

mmaaz-git replied:
I see, that makes sense. I definitely think we need to handle multiplicity properly. I will work on the square-free stuff.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -6744,17 +6810,15 @@ def all_roots(f, multiple=True, radicals=True):
-    ...
-    NotImplementedError: sorted roots not supported over EX
+    >>> all_roots(p, extension=True)
+    [sqrt(2), sqrt(2), sqrt(3), sqrt(3)]
```
I think this is good.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -6831,6 +6899,9 @@ def all_roots(f, multiple=True, radicals=True):
     """
     try:
         F = Poly(f, greedy=False)
+        if extension and not F.domain.is_AlgebraicField:
+            F = Poly(F.expr, extension=True)
```
Do we need `greedy=False` here?

The effect of greedy is that it means that numbers like `sqrt(2)` are considered coefficients rather than generators:
```python
In [11]: Poly(x + sqrt(2))
Out[11]: Poly(x + (sqrt(2)), x, sqrt(2), domain='ZZ')

In [12]: Poly(x + sqrt(2), greedy=False)
Out[12]: Poly(x + sqrt(2), x, domain='EX')
```
I don't think that matters here though because we should reject multivariate polynomials and also polynomials whose domain is not QQ or (if extension=True) AlgebraicField.

I think this should just be:
```python
F = poly(f, extension=extension)
```

mmaaz-git replied:
I just tried this but it causes some conflict in polyoptions. 

```
>>> Poly(x + sqrt(2), x, extension=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/maaz/Documents/sympy/sympy/polys/polytools.py", line 168, in __new__
    opt = options.build_options(gens, args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/polyoptions.py", line 742, in build_options
    return Options(gens, args)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/polyoptions.py", line 153, in __init__
    preprocess_options(args)
  File "/Users/maaz/Documents/sympy/sympy/polys/polyoptions.py", line 151, in preprocess_options
    self[option] = cls.preprocess(value)
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/polyoptions.py", line 552, in preprocess
    raise OptionError("'False' is an invalid argument for 'extension'")
sympy.polys.polyerrors.OptionError: 'False' is an invalid argument for 'extension'
```

In light of that, I think it makes sense to keep the `greedy=False` option which is as before?

oscarbenjamin replied:
The polyoptions stuff is a bit over-engineered and complicates simple things. I guess we can keep greedy. We shouldn't construct the Poly twice though.

mmaaz-git replied:
Sure

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,13 +775,50 @@ def _get_roots(cls, method, poly, radicals):
+                poly = poly.replace(d, x)
                 break
+
+        if dom.is_QQ or dom.is_ZZ:
+            return cls._get_roots_qq(method, poly, radicals)
+        if dom.is_AlgebraicField:
+            return cls._get_roots_alg(method, poly, radicals)
+        else:
```
Gaussian domains also need to be handled here:
```python
In [19]: all_roots(x**2 + I, extension=True)
---------------------------------------------------------------------------
NotImplementedError 
```

mmaaz-git replied:
Done!

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3816,72 @@ def nth_power_roots_poly(f, n):
+        >>> f.which_roots([-1, 1, -I, I, 0], real=False)
+        [-1, 1, -I, I]
+        >>> f.which_roots([2, 5], real=True) # expectedly wrong
+        [2, 5]
```
Rather than showing this example I think it might be clearer to show a demonstration involving `lift`.

Also it would be good to have:
```
See Also
=======

same_root
```
Likewise `same_root` could reference `which_roots`.

mmaaz-git replied:
Sounds good, will do

oscarbenjamin commented:
I think this is basically right now apart from minor things.

mmaaz-git committed:
Added tests for real_roots and all_roots with extension

mmaaz-git committed:
Changed implementation of which_root() to use Counter

mmaaz-git committed:
polytools: New method root_multiplicity()

The new Poly.root_multiplicity() method finds the multiplicity
of a root with numerical evaluation of its derivatives. This is
necessary for the which_roots() method to work properly for
real roots, as count_roots() returns the count of unique roots.
This meant that the loop in which_roots() would not terminate,
until an overflow error. This was previously breaking some
examples in real_roots(). Therefore, the loop condition was
changed, and multiplicity computed after.

Examples of errors before the change:

>>> Poly((x-1)**2, x).which_roots([1, 1])
OverflowError: cannot convert float infinity to integer

>>> Poly((x-sqrt(2))**2, x, extension=True).real_roots()
OverflowError: cannot convert float infinity to integer

In the last example, observe the roots of the lift:

>>> Poly((x-sqrt(2))**2, x, extension=True).lift().real_roots()
[-sqrt(2), -sqrt(2), sqrt(2), sqrt(2)]

With the change, this works fine:

>>> Poly((x-sqrt(2))**2, x, extension=True).real_roots()
[sqrt(2), sqrt(2)]

mmaaz-git committed:
Changed test in which_roots()

mmaaz-git committed:
Added tests for root_multiplicity

mmaaz-git committed:
Minor change for flake

mmaaz-git committed:
Don't construct poly twice in real_roots and all_roots

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3817,130 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def root_multiplicity(f, root, prec=10):
+        """
+        Computes the multiplicity of a root for f, at a given
+        numerical precision. Works by evaluating derivatives of
+        f at the given root. If it is not a root of f (up to
+        numerical precision), returns zero.
```
I don't think that this function is needed in this context because we will already factorise the polynomial into irreducibles or at least square free parts so the multiplicity is already handled.

Also factorisation is a better method than numerical evaluation for this. Only exact factorisation can reveal multiple roots reliably.

mmaaz-git committed:
Fixed docstring title underline

mmaaz-git committed:
Docstring typo

mmaaz-git committed:
rootoftools: _get_roots_alg now uses square-free factorization

Implementation of _get_roots_alg() now first computes the square-
free factorization, finds the roots of each factor, and then puts
them together and sorts. This is to avoid issues with multiplicity
of roots, which could arise when calling Poly.which_roots(). Also,
Poly.which_roots() now returns unique elements, which is useful. Added
a new method to ComplexRootOf called _sort_roots() which sorts a list
of roots in the usual order: reals first, then complexes by the real
and imaginary part, with conjugates together. Also removed the method
Poly.root_multiplicity() as it is not needed anymore and was also
not a good implementation. Lastly, added tests for polynomials with
algebraic coefficients with multiple roots.

mmaaz-git committed:
Changed solve test back as extension is False by default

mmaaz-git committed:
rootoftools: Changed _sort_roots() method

Previously, _sort_roots() assumed that complex roots appeared
in conjugate pairs. However, this is not true if the polynomial
in question is over a Gaussian domain (e.g. x-I). So, had to
change the way it is sorting the roots. Then, an issue arose
where some roots are not directly comparable, and SymPy
would throw a TypeError. Therefore, I now try to compare roots
symbolically if possible, but in the case of TypeError, fall back
on numerical evaluation. This is not ideal but I think it is the
best we can do without having information like root isolating
intervals.

mmaaz-git committed:
rootoftools: get_roots() now handles Gaussian domains

get_roots() now checks for Gaussian domains ZZ_I or QQ_I and then
dispatches to get_roots_alg(). This enables finding roots of, e.g.,
x**2 + I. This also works for, e.g., x**2 + sqrt(2)*I. Added a
unit test for this, as well as added a little statement about this
under all_roots() in polytools. Note that this only works when
calling all_roots() -- an error will be thrown if real_roots() is
called on such polynomials as root-counting is not supported over
this domain (although in the examples outlined there are no real
roots at all).

mmaaz-git committed:
Minor typos

mmaaz-git commented:
I have changed _get_roots_alg to use square-free factorization first. One concern was how to get the roots properly sorted. Currently, the methods in CRootOf, e.g., when solving over rational coeffs, store isolating intervals and use these to sort, and then finally convert these to radicals or CRootOf objects at the very end. There are sorting methods currently in CRootOf that accomplish this, but they require this intermediary data that CRootOf uses. However, when I tried to recreate this pattern for the square free factors, I couldn't see a way to somehow also do the which_roots() procedure in between. The inputs/outputs of the various internal CRootOf methods have some particular formats. 

All that said, I ended up going for: compute square-free factors, get roots of each factor, apply which_roots, and then merge them all at the end, and then sort. I wrote a new _sort_roots() procedure in CRootOf which sorts roots in the usual way.

I also changed `get_roots()` so that it can detect Gaussian domains and dispatch accordingly to `_get_roots_alg()`. I made other additions to unit tests and doctests.

mmaaz-git committed:
retrigger checks

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -6744,17 +6827,23 @@ def all_roots(f, multiple=True, radicals=True):
     >>> [r.evalf(3) for r in all_roots(p)]
     [1.17, -0.765 - 0.352*I, -0.765 + 0.352*I, 0.181 - 1.08*I, 0.181 + 1.08*I]
 
-    Irrational algebraic or transcendental coefficients cannot currently be
-    handled by :func:`all_roots` (or :func:`~.rootof` more generally):
+    Irrational algebraic coefficients are handled by :func:`all_roots`
+    if `extension=True` is set.
 
     >>> from sympy import sqrt, expand
     >>> p = expand((x - sqrt(2))*(x - sqrt(3)))
     >>> print(p)
     x**2 - sqrt(3)*x - sqrt(2)*x + sqrt(6)
-    >>> all_roots(p)
-    Traceback (most recent call last):
-    ...
-    NotImplementedError: sorted roots not supported over EX
+    >>> all_roots(p, extension=True)
+    [sqrt(2), sqrt(3)]
```
I think it is good to show both examples here i.e. what happens with and without `extension=True`.

mmaaz-git replied:
Done

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -407,6 +408,33 @@ def _eval_is_imaginary(self):
+        """
+        real_roots = sorted(r for r in roots if r.is_real)
+
+        try:
+            complex_roots = sorted((r for r in roots if not r.is_real),
+                                key=lambda root: (re(root), im(root)))
+            """
+            Example error
+            TypeError: cannot determine truth value of Relational:
+            re(CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 3)) <
+                re(CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 2))
+            """
+        except TypeError:
+            complex_roots = sorted((r for r in roots if not r.is_real),
+                                key=lambda root: (re(root.evalf()), im(root.evalf())))
```
I don't think that this fallback should be used. It cannot reliably order the roots. If we get here then it is almost certainly because two or more roots have identical real part but calling `re(root.eval())` will make them seem not equal leading to an arbitrary order.

Is it not possible to continue using the existing code that sorts the roots?

I realise everything here is that the code here is quite convoluted so making it difficult to do these things.

Here is one way to do this:
```python
roots = p.lift().real_roots() # Existing QQ code can sort the roots

subroots = set()
for f, m in p.sqf_list()[1]:
    subroots.update(p.which_roots(roots, real=True))

roots_filtered = [r for r in roots if r not in subroots]
```
We only need to make this square-free for the real_roots case and only because otherwise count_roots does not give us the multiplicities. For the all_roots case it is just:
```
roots_filtered = p.which_roots(p.lift().all_roots(), real=False)
```
Given that functions like `real_roots` and `all_roots` are separate functions rather than being separated by a keyword argument I think it would make sense to have `which_real_roots_sqf` and `which_roots` as separate methods. The `which_real_roots_sqf` function requires the polynomial to be square free. The `which_roots` function does not. A function `which_real_roots()` could do the dance with `sqf_list` that I showed above before calling `which_real_roots_sqf()`.

mmaaz-git replied:
This is great! Thanks for the help.

After some experimenting, I think that even in the all_roots case, we need to get square-free factors. Eg if `f= Poly((x-1) * (x-sqrt(2))**2 * (x-I) * (x+I), x, extension=True)` (which has roots `[1,sqrt(2),sqrt(2), I, -I]`).

Bu it's lift has roots `[-sqrt(2), -sqrt(2), 1, 1, sqrt(2), sqrt(2), -I, -I, I, I]`. The `-sqrt(2)` are filtered out easily. However, in `which_roots`, if we just compare the sum of counts to the degree (as in the all_roots setting), this will loop forever. Thus, I think it makes sense to always return unique roots in `which_roots()`. And then handle multiplicity at the level of the square-free factorization.

In light of that, I think both cases are similar enough that we can keep `which_roots` one function, separated by the `real` argument? The only difference is the root counting. Otherwise, both should return unique elements.

oscarbenjamin replied:
> I think that even in the all_roots case, we need to get square-free factors.

Yes, you are right. The multiplicities might not be equal so we can't count the roots with multiplicities except in the square free case where the multiplicity is always 1.

mmaaz-git committed:
Changed docstrings of real_roots() and all_roots()

mmaaz-git committed:
rootoftools: Changed implementation of _get_roots_alg

_get_roots_alg() now first calls _get_roots_qq on the lift to
get sorted polynomials. It then calls which_roots on each square
free factor, in order to get the correct multiplicity. The method
ComplexRootOf._sort_roots() is removed.

mmaaz-git commented on sympy/polys/polytools.py:
```diff
@@ -6744,17 +6827,23 @@ def all_roots(f, multiple=True, radicals=True):
-    Irrational algebraic or transcendental coefficients cannot currently be
-    handled by :func:`all_roots` (or :func:`~.rootof` more generally):
+    Irrational algebraic coefficients are handled by :func:`all_roots`
+    if `extension=True` is set.
```
Done

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,64 @@ def _get_roots(cls, method, poly, radicals):
+                roots_filt = f.which_roots(roots, real=True)
+            elif method == "_all_roots":
+                roots_filt = f.which_roots(roots, real=False)
+            for r in roots_filt:
+                if r in subroots:
+                    subroots[r] += m
+                else:
+                    subroots[r] = m
```
We can write this as just `subroots[r] = m`.

Is it possible for distinct square-free factors to have any roots in common?

I don't think it is. Let's suppose that `fi` and `fj` are factors with multiplicity `mi` and `mj`. If some number `r` is a root of both `fi` and `fj` then it must be a root of some `fij` that is a factor of both `fi` and `fj` and then the multiplicity of `fij` would be at least `mi + mj`.

To put it another way the distinct square-free factorisation is defined so that the factors are pairwise coprime (the algorithm literally divides out the gcds). That means that they cannot have any roots in common.

mmaaz-git replied:
Ah great point, I put that in case, but from your explanation I can see it wouldn't happen.

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -772,15 +775,64 @@ def _get_roots(cls, method, poly, radicals):
+                else:
+                    subroots[r] = m
+
+        # loop over unique roots, keep ones in subroots, and store multiplicity
+        roots_filtered_mult = [(r, subroots[r]) for r in list(dict.fromkeys(roots)) if r in subroots]
+
+        # flatten the roots list according to multiplicity
+        roots_flat = sum(([r] * m for r, m in roots_filtered_mult), [])
```
We can make this simpler:
```python
roots_flat = []
for r in roots:
    if r in subroots:
        m = subroots[r]
        roots_flat.extend([r] * m)
```

mmaaz-git replied:
Sounds good, although we have to add a way to not add the same root again to roots_flat. Eg with this code we get
```
In [5]: Poly((x-1) * (x-sqrt(2))**2, x, extension=True).real_roots()
Out[5]: [1, 1, sqrt(2), sqrt(2), sqrt(2), sqrt(2)]
```
Bc the lift has roots `[-sqrt(2), -sqrt(2), 1, 1, sqrt(2), sqrt(2)]`, so the sqrt(2) gets looped over twice (and mult is 2, so we get 4 of them in the final output). 

I've added a check for this to your code.

```python
        roots_seen = set()
        roots_flat = []
        for r in roots:
            if r in subroots and r not in roots_seen:
                m = subroots[r]
                roots_flat.extend([r] * m)
                roots_seen.add(r)
```

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -7003,7 +7123,17 @@ def real_roots(f, multiple=True, radicals=True):
     .. [1] https://en.wikipedia.org/wiki/Casus_irreducibilis
     """
     try:
-        F = Poly(f, greedy=False)
+        if isinstance(f, Poly):
+            if extension and not f.domain.is_AlgebraicField:
+                F = Poly(f.expr, extension=True)
+            else:
+                F = f
+        else:
+            if extension:
+                F = Poly(f, extension=True)
+            else:
+                F = Poly(f, greedy=False)
```
There should be a simpler way of writing this. One problem with the above is that we call `Poly(f.expr, extension=True)` if we have a Poly with domain QQ which is unnecessary.

It seems like something is wrong if we can't do all of this in one call. But I can't think what the call should be...

mmaaz-git replied:
One reason I have all these checks is because if we just did something like:
```python
if extension:
            F = Poly(f, extension=True)
else:
            F = Poly(f, greedy=False)
```

the issue is that if specify extension=True if the Poly is already made, it won't change it. E.g.

```
In [3]: f = x - sqrt(2)

In [4]: f
Out[4]: x - sqrt(2)

In [5]: Poly(f, x)
Out[5]: Poly(x - sqrt(2), x, domain='EX')

In [6]: Poly(f, x, extension=True)
Out[6]: Poly(x - sqrt(2), x, domain='QQ<sqrt(2)>')

In [7]: Poly(Poly(f, x), extension=True)
Out[7]: Poly(x - sqrt(2), x, domain='EX')

In [8]: Poly(Poly(f, x).expr, extension=True)
Out[8]: Poly(x - sqrt(2), x, domain='QQ<sqrt(2)>')
```

But as the last output shows, if you use the expr then you can construct the extension. It seems important to me to handle the case that extension=True and the user passes a Poly. I'm not sure if there is a better way to construct the domain in this case.

oscarbenjamin replied:
Maybe if the input is a Poly it is reasonable to just ignore the other options like extension=True.

Either way I was more complaining that there isn't a simple way to do what is being done here that does not require handling 4 separate cases.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3817,89 @@ def nth_power_roots_poly(f, n):
+        >>> f.which_roots([2, 5], real=True)
+        [2, 5]
+
+        # lifting to rational coeffs produces extraneous roots
+        # which we can filter out with this method
+        >>> f = Poly(sqrt(2)*x**3 + x**2 - 1, x, extension=True)
+        >>> f.lift()
```
The lines that look like comments here should not be like comments. This is prose in the docs. We don't have comments in the docs.

mmaaz-git replied:
Changed to prose!

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3817,89 @@ def nth_power_roots_poly(f, n):
+
+        # expected behaviour if not a superset of roots
+        >>> f.which_roots([2, 5], real=True)
+        [2, 5]
```
I don't think we want to call this "expected behaviour". We just need to make clear what the premise of the function is and why it is useful to have a function with that premise.

oscarbenjamin replied:
The example with lift makes that clear.

mmaaz-git replied:
Sounds good. I've moved this out of the doc but I put this case where candidates is not a superset into a unit test as I think it's a good edge case to cover.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3817,89 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def which_roots(f, candidates, real=True):
+        """
+        Given a superset of roots of f, finds which ones are
+        roots of f. Note this won't work properly if candidates is
+        not a superset of the roots of f. The list returned has
+        unique elements, is at most the length of the number of
+        unique roots of f, and preserves the order of the original
+        list of candidates. The `real` argument only controls
+        the root counting: if `real=True`, the `.count_roots()` method
+        is used (only supported over ZZ, QQ, or AlgebraicField); if
+        `real=False`, the degree is used.
```
A docstring should start with a single line summary:

https://docs.sympy.org/latest/contributing/docstring.html

I would write this as:
```python
    def which_roots(f, candidates, real=True):
        """
        Find roots of a square-free polynomial ``f`` from ``candidates``.

        Explanation
        ===========

        If ``f`` is a square-free polynomial and ``candidates`` is a superset
        of the roots of ``f``, then ``f.which_roots(candidates)`` returns a
        list containing exactly the set of roots of ``f``. The domain must be
        :ref:`ZZ`, :ref:`QQ`, :ref:`ZZ_I`, :ref:`QQ_I` or :ref:`QQ(a)` and
        ``f`` must be univariate and square-free.

        If ``real=True`` then ``candidates`` must be a superset of the real
        roots of ``f`` and ``f.which_roots(candidates)`` returns the set of
        real roots of ``f``. Otherwise if ``real=False`` then ``candidates``
        must be a superset of all complex roots of ``f`` and
        ``f.which_roots(candidates, real=False)`` returns exactly the set of
        all complex roots of ``f``.

        """
```
I still think though that it would be better not to have the `real_roots=True/False` keyword argument. Rather it should just be two different functions like `which_real_roots` and `which_all_roots` or something.

mmaaz-git replied:
Thanks for writing this! I've also separated it out into `which_real_roots` and `which_all_roots`.

mmaaz-git committed:
Simplify storing subroots counts

mmaaz-git committed:
Simplify making final list of roots

mmaaz-git committed:
polytools: Split which_roots into methods for real and all

Split Poly.which_roots() into Poly.which_real_roots() and
Poly.which_all_roots(). Added new documentation for them about
the requirement for the polynomial to be square free. Added new
tests to test edge cases where candidates is not a superset or when
the polynomial is not square free. Lastly, changed the calls in
CRootOf._get_roots_alg() to these two methods.

mmaaz-git committed:
Remove unnecessary import

mmaaz-git commented:
Hmm looks like the PR is failing bc of Ruff (mentioned in #26964)

mmaaz-git committed:
Change reference to which_roots in docs

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3816,6 +3817,166 @@ def nth_power_roots_poly(f, n):
+        # using Counter bc its like an ordered set
+        root_counts = Counter(candidates)
+
+        while len(root_counts) > num_roots:
+            for r in list(root_counts.keys()):
+                # If f(r) != 0 then f(r).evalf() gives a float/complex with precision.
+                f_r = f(r).evalf(prec, maxn=2*prec)
+                if abs(f_r)._prec >= 2:
+                    root_counts.pop(r)
```
A lot of the code in `which_real_roots` and `which_all_roots` is duplicated. It would be better factor it out into a separate method that could be called by both public methods.

oscarbenjamin replied:
Apart from this it all looks good though so let's fix this and get it merged.

I suggest having a private method like
```
def _which_roots(self, num_roots, candidates):
   ...
```
Then both `which_all_roots` and `which_real_roots` can call that method and it can house all the otherwise repeated code.

mmaaz-git replied:
Done!

mmaaz-git committed:
Wrote internal function _which_roots

mmaaz-git commented:
I changed the release notes entry
"Added ability to compute roots of polynomials with real algebraic coefficients"
to
"Added ability to compute roots of polynomials with algebraic coefficients"

since we also handle gaussian domains now

oscarbenjamin commented:
Okay, looks good. Thanks!

oscarbenjamin commented:
This is a nice addition and I think it was good that we went through the full review process on a smaller PR so that there is a clear idea of what it is that is needed to the final product into the codebase.

Looking forward to seeing the next steps for CAD!

mmaaz-git commented:
Thanks! I learned a lot during the process :)
